libraries/AP_HAL_Linux/PWM_Sysfs.cpp

```diff
@@ -0,0 +1,168 @@
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#include <AP_HAL/AP_HAL.h>
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_LINUX
+
+#include "PWM_Sysfs.h"
+
+#include <fcntl.h>
+#include <errno.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+using namespace Linux;
+
+static const AP_HAL::HAL& hal = AP_HAL::get_HAL();
+
+#define BASE_PWM_PATH "/sys/class/pwm/pwmchip%d"
+
+#define UINT32_STR_MAX 10 + 1
+
+PWM_Sysfs::PWM_Sysfs(uint8_t chip, uint8_t channel)
+    : _chip(chip)
+    , _channel(channel)
+    , _duty_cycle_fd(-1)
+    , _nsec_duty_cycle_value(0)
+{
+    char path[PATH_MAX];
+    Util* util = Util::from(hal.util);
+
+    snprintf(path, sizeof(path), BASE_PWM_PATH "/export", _chip);
+    if (util->write_file(path, "%u", _channel) < 1) {
+        hal.scheduler->panic(PSTR("LinuxPWM_Sysfs: Unable to export PWM pin."));
+    }
+
+    snprintf(path, sizeof(path), BASE_PWM_PATH "/pwm%d/duty_cycle", _chip, _channel);
```

Comment:
Huum, so I will replace PATH_MAX by:
#define PWM_PATH_MAX sizeof("/sys/class/pwm/pwmchipXXX/pwmXXX/duty_cycle")
'duty_cycle' because it is the bigger name